traits/trait_list_object.py

```diff
@@ -53,465 +90,765 @@ def __init__(self, index=0, removed=None, added=None):
-    appropriate index, added and removed values.  In the case of setting
-    or deleting items with a slice, the index will hold either:
-
-    - the location of the first element changed, if the step is 1.  In
-      this case the last element changed can be inferred from the
-      length of the removed values.
-    - the slice that was used, otherwise.
+class TraitList(list):
+    """ A subclass of list that validates and notifies listeners of changes.
 
     Parameters
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : HasTraits instance
-        The HasTraits instance that the list has been set as an attribute for.
-    name : str
-        The name of the attribute on the object.
     value : list
-        The list of values to initialize the TraitListObject with.
+        The value for the list
+    validator : callable, optional
+        Called to validate items in the list.
+        With the signature::
+
+            validator(current_list, index, removed, added)
+
+        If the validator is not provided, all values added to the list
+        are considered to be valid.
+
+    notifiers : list of callable, optional
+        A list of callables with the signature::
+
+            notifier(trait_list, index, removed, added)
+
+        If a notifier is not provided, an empty list of notifiers is
+        used.
 
     Attributes
     ----------
-    trait : CTrait instance
-        The CTrait instance associated with the attribute that this list
-        has been set to.
-    object : weak reference to a HasTraits instance
-        A weak reference to a HasTraits instance that the list has been set
-        as an attribute for.
-    name : str
-        The name of the attribute on the object.
-    name_items : str
-        The name of the items event trait that the trait list will fire when
-        mutated.
+    validator : callable, optional
+        Called to validate items in the list
+        With the signature::
+
+            validator(current_list, index, removed, added)
+
+        If the validator is not provided, all values added to the list
+        are considered to be valid.
+
+    notifiers : list of callable, optional
+        A list of callables with the signature::
+
+            notifier(trait_list, index, removed, added)
+
+        If a notifier is not provided, an empty list of notifiers is
+        used.
+
     """
 
-    def __init__(self, trait, object, name, value):
-        self.trait = trait
-        self.object = ref(object)
-        self.name = name
-        self.name_items = None
-        if trait.has_items:
-            self.name_items = name + "_items"
+    def __new__(cls, *args, **kwargs):
+        instance = super().__new__(cls)
+        instance.validator = None
+        instance.notifiers = []
+        return instance
+
+    def __init__(self, value=(), *, validator=None, notifiers=()):
```

Comment:
Please make the default here `None` rather than the empty tuple (`notifiers` is a list, not a tuple).